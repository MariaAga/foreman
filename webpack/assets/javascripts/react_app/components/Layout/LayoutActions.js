import {
  LAYOUT_INITIALIZE,
  LAYOUT_SHOW_LOADING,
  LAYOUT_HIDE_LOADING,
  LAYOUT_CHANGE_LOCATION,
  LAYOUT_CHANGE_ACTIVE,
  LAYOUT_CHANGE_ORG,
  LAYOUT_CHANGE_IS_NAV_OPEN,
} from './LayoutConstants';

export const initializeLayout = ({
  items,
  activeMenu,
  organization,
  location,
}) => ({
  type: LAYOUT_INITIALIZE,
  payload: {
    items,
    activeMenu,
    organization,
    location,
  },
});

export const showLoading = () => ({
  type: LAYOUT_SHOW_LOADING,
});

export const hideLoading = () => ({
  type: LAYOUT_HIDE_LOADING,
});

export const changeActiveMenu = ({ title }) => ({
  type: LAYOUT_CHANGE_ACTIVE,
  payload: {
    activeMenu: title,
  },
});

export const changeOrganization = organization => ({
  type: LAYOUT_CHANGE_ORG,
  payload: { organization },
});

export const changeLocation = location => ({
  type: LAYOUT_CHANGE_LOCATION,
  payload: {
    location,
  },
});

export const changeIsNavOpen = isNavOpen => ({
  type: LAYOUT_CHANGE_IS_NAV_OPEN,
  payload: {
    isNavOpen,
  },
});
