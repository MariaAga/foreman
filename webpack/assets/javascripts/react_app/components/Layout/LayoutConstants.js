export const LAYOUT_INITIALIZE = 'LAYOUT_INITIALIZE';
export const LAYOUT_SHOW_LOADING = 'LAYOUT_SHOW_LOADING';
export const LAYOUT_HIDE_LOADING = 'LAYOUT_HIDE_LOADING';
export const LAYOUT_CHANGE_IS_NAV_OPEN = 'LAYOUT_CHANGE_IS_NAV_OPEN';
