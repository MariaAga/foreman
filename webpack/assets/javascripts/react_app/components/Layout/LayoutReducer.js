import Immutable from 'seamless-immutable';

import {
  LAYOUT_INITIALIZE,
  LAYOUT_SHOW_LOADING,
  LAYOUT_HIDE_LOADING,
  LAYOUT_CHANGE_IS_NAV_OPEN,
} from './LayoutConstants';

const initialState = Immutable({
  items: [],
  isLoading: false,
  isNavOpen: true,
});

export default (state = initialState, action) => {
  const { payload, type } = action;

  switch (type) {
    case LAYOUT_INITIALIZE:
      return state
        .set('items', payload.items)
        .set('isNavOpen', payload.isNavOpen)
        .set('currentOrganization', payload.organization)
        .set('currentLocation', payload.location);

    case LAYOUT_SHOW_LOADING:
      return state.set('isLoading', true);

    case LAYOUT_HIDE_LOADING:
      return state.set('isLoading', false);

    case LAYOUT_CHANGE_IS_NAV_OPEN:
      return state.set('isNavOpen', payload.isNavOpen);

    default:
      return state;
  }
};
