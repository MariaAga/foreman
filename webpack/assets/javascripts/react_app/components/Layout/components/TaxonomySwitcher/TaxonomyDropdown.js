import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { ContextSelectorItem } from '@patternfly/react-core';
import CustomContextSelector from '../CustomContextSelector';
import { noop } from '../../../../common/helpers';
import { translate as __ } from '../../../../common/I18n';
import './TaxonomyDropdown.scss';

const TaxonomyDropdown = ({
  taxonomyType,
  currentTaxonomy,
  taxonomies,
  changeTaxonomy,
  anyTaxonomyText,
  manageTaxonomyText,
  anyTaxonomyURL,
  manageTaxonomyURL,
  ...props
}) => {
  const [searchValue, setSearchValue] = useState('');
  const [isOpen, setIsOpen] = useState(false);
  const [filteredItems, setFilteredItems] = useState(taxonomies);

  const onToggle = (event, newIsOpen) => {
    setIsOpen(newIsOpen);
  };
  const onSelect = () => {
    setIsOpen(!isOpen);
  };
  const onSearchInputChange = (value, event) => {
    setSearchValue(event.target.value);
    onSearchButtonClick();
  };
  const onSearchButtonClick = () => {
    const filtered =
      searchValue === ''
        ? taxonomies
        : taxonomies.filter(item =>
            item.title.toLowerCase().includes(searchValue.toLowerCase())
          );
    setFilteredItems(filtered || []);
  };

  const staticGroup = {
    title: __(taxonomyType),
    items: [
      {
        title: __(anyTaxonomyText),
        href: anyTaxonomyURL,
        onClick: () => {
          changeTaxonomy({ title: anyTaxonomyText });
        },
        className: `${taxonomyType.toLowerCase()}s_clear`,
      },
      {
        title: __(manageTaxonomyText),
        href: manageTaxonomyURL,
        className: taxonomyType.toLowerCase(),
      },
    ],
  };
  return (
    <CustomContextSelector
      toggleText={__(currentTaxonomy)}
      onSearchInputChange={onSearchInputChange}
      isOpen={isOpen}
      searchInputValue={searchValue}
      onToggle={onToggle}
      onSelect={onSelect}
      onSearchButtonClick={onSearchButtonClick}
      screenReaderLabel="Selected Taxonomy:"
      showFilter={taxonomies.length > 6}
      searchProps={{
        className: 'taxonomy_search',
        id: `search_taxonomy_${taxonomyType.toLowerCase()}`,
      }}
      staticGroup={staticGroup}
      {...props}
    >
      {filteredItems.map((taxonomy, i) => (
        <ContextSelectorItem key={i}>
          <a
            id={`select_taxonomy_${taxonomy.title}`}
            className={`${taxonomyType.toLowerCase()}_menuitem`}
            href={taxonomy.href}
            onClick={() => {
              changeTaxonomy({ title: taxonomy.title, id: taxonomy.id });
            }}
          >
            {__(taxonomy.title)}
          </a>
        </ContextSelectorItem>
      ))}
    </CustomContextSelector>
  );
};

TaxonomyDropdown.propTypes = {
  taxonomyType: PropTypes.string.isRequired,
  currentTaxonomy: PropTypes.string.isRequired,
  taxonomies: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string,
      href: PropTypes.string.isRequired,
    })
  ).isRequired,
  id: PropTypes.string.isRequired,
  changeTaxonomy: PropTypes.func,
  anyTaxonomyText: PropTypes.string.isRequired,
  manageTaxonomyText: PropTypes.string.isRequired,
  anyTaxonomyURL: PropTypes.string.isRequired,
  manageTaxonomyURL: PropTypes.string.isRequired,
};

TaxonomyDropdown.defaultProps = {
  changeTaxonomy: noop,
};

export default TaxonomyDropdown;
