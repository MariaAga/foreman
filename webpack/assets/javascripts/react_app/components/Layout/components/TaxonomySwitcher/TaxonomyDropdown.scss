.pf-c-context-selector__menu-list-item a.organization_menuitem,
.pf-c-context-selector__menu-list-item a.location_menuitem {
  text-decoration: inherit;
  color: inherit;
}
