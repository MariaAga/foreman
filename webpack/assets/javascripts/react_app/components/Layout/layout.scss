@import '../../common/colors.scss';

.secondary-nav-item-pf > .nav-pf-secondary-nav {
  background-color: $nav-pf-vertical-active-bg-color;

  .mobile-active {
    // Mark active Location/Org in Mobile view
    a {
      .list-group-item-value {
        background-color: $nav-pf-vertical-secondary-active-bg-color;
      }
    }
  }

  .list-group-item {
    a {
      background-color: $nav-pf-vertical-active-bg-color;

      &:hover {
        .list-group-item-value {
          background-color: $nav-pf-vertical-secondary-active-bg-color;
        }
      }
    }

    span {
      color: $nav-pf-vertical-active-icon-color;
    }
  }
}

#react-app-root > .nav-pf-vertical {
  background-color: $nav-pf-vertical-secondary-active-bg-color;

  .list-group {
    .list-group-item {
      a {
        span {
          color: $nav-pf-vertical-icon-color;
        }

        &:hover {
          background-color: $nav-pf-vertical-active-bg-color;

          span {
            color: $nav-pf-vertical-active-icon-color;
          }
        }

        &::after {
          color: $nav-pf-vertical-icon-color;
        }
      }

      &:hover {
        a {
          background-color: $nav-pf-vertical-active-bg-color;

          span {
            color: $nav-pf-vertical-active-icon-color;
          }
        }
      }

      &.active {
        a {
          background-color: $nav-pf-vertical-active-bg-color;

          span {
            color: $nav-pf-vertical-active-icon-color;
          }
        }
      }
    }
  }

  .visible-xs-block {
    display: none !important;

    @media (max-width: 1002px) {
      display: block !important;
    }
  }
}

#account_menu {
  outline: none;
}

#vertical-spinner {
  margin-top: 15px;
  margin-left: 5px;
}

.navbar-header {
  width: 42px;
}

#navbar-header {
  left: 42px;
}

#data-toolbar {
  padding: 0;
  background-color: transparent;
}
