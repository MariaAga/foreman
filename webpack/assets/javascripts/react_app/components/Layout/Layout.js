import React from 'react';
import {
  Brand,
  Page,
  PageSection,
  PageSectionVariants,
  PageSidebar,
  SkipToContent,
  PageHeader,
} from '@patternfly/react-core';

import { translate as __ } from '../../common/I18n';
import VerticalNav from './components/VerticalNav';
import HeaderToolbar from './components/Toolbar/HeaderToolbar';

import { layoutPropTypes, layoutDefaultProps } from './LayoutHelper';
import './layout.scss';

const Layout = ({
  items,
  data,
  isLoading,
  isNavOpen,
  navigate,
  children,
  changeIsNavOpen,
}) => {
  const onNavToggle = () => {
    isNavOpen ? changeIsNavOpen(false) : changeIsNavOpen(true);
  };
  const header = (
    <PageHeader
      logo={
        <React.Fragment>
          <Brand src={data.logo} alt={data.brand} href={data.root} />
          <span className="navbar-brand-txt">
            <span>{data.brand}</span>
          </span>
        </React.Fragment>
      }
      logoProps={{ href: data.root }}
      showNavToggle
      isManagedSidebar
      onNavToggle={onNavToggle}
      headerTools={<HeaderToolbar {...data} isLoading={isLoading} />}
    />
  );

  const sidebar = (
    <PageSidebar
      isNavOpen={isNavOpen}
      nav={<VerticalNav items={items} navigate={navigate} />}
    />
  );
  const pageId = 'main';
  const pageSkipToContent = (
    <SkipToContent href={`#${pageId}`}>{__('Skip to content')}</SkipToContent>
  );
  return (
    <Page
      header={header}
      sidebar={sidebar}
      skipToContent={pageSkipToContent}
      mainContainerId={pageId}
    >
      <PageSection
        className="react-container nav-pf-persistent-secondary"
        variant={PageSectionVariants.light}
      >
        {children}
      </PageSection>
    </Page>
  );
};
Layout.propTypes = layoutPropTypes;

Layout.defaultProps = layoutDefaultProps;

export default Layout;
