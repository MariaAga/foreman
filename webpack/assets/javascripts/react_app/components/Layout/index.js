import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useHistory } from 'react-router-dom';

import {
  initializeLayout,
  collapseLayoutMenus,
  expandLayoutMenus,
  changeIsNavOpen,
} from './LayoutActions';
import reducer from './LayoutReducer';
import {
  patternflyMenuItemsSelector,
  selectIsLoading,
  selectIsNavOpen,
} from './LayoutSelectors';
import { combineMenuItems } from './LayoutHelper';

import { getIsNavbarOpen } from './LayoutSessionStorage';

import Layout from './Layout';

const ConnectedLayout = ({ children, data }) => {
  const dispatch = useDispatch();
  useEffect(() => {
    dispatch(
      initializeLayout({
        items: combineMenuItems(data),
        isNavOpen: getIsNavbarOpen(),
        organization: data.orgs.current_org,
        location: data.locations.current_location,
      })
    );
  }, [data, dispatch]);

  const isNavOpen = useSelector(state => selectIsNavOpen(state));
  useEffect(() => {
    // toggles a class in the body tag, so that the main #rails-app-content container can have the appropriate width
    if (isNavOpen) {
      document.body.classList.add('pf-m-expanded');
    } else {
      document.body.classList.remove('pf-m-expanded');
    }
  }, [isNavOpen]);

  const { push: navigate } = useHistory();
  const items = useSelector(state => patternflyMenuItemsSelector(state));
  const isLoading = useSelector(state => selectIsLoading(state));
  return (
    <Layout
      data={data}
      navigate={navigate}
      items={items}
      isLoading={isLoading}
      isNavOpen={isNavOpen}
      collapseLayoutMenus={() => dispatch(collapseLayoutMenus())}
      expandLayoutMenus={() => dispatch(expandLayoutMenus())}
      changeIsNavOpen={value => dispatch(changeIsNavOpen(value))}
    >
      {children}
    </Layout>
  );
};

// export prop-types
export const { propTypes, defaultProps } = Layout;

ConnectedLayout.propTypes = {
  children: propTypes.children,
  data: propTypes.data,
};

ConnectedLayout.defaultProps = {
  children: defaultProps.children,
  data: defaultProps.data,
};

// export reducers
export const reducers = { layout: reducer };

// export connected component
export default ConnectedLayout;
