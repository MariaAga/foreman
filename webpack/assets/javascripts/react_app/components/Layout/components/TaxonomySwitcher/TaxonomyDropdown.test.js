import { testComponentSnapshotsWithFixtures } from '../../../../common/testHelpers';
import TaxonomyDropdown from './TaxonomyDropdown';
import { hasTaxonomiesMock } from '../../Layout.fixtures';

const props = {
  taxonomyType: 'Organization',
  id: 'organization-dropdown',
  currentTaxonomy: hasTaxonomiesMock.currentOrganization,
  taxonomies: hasTaxonomiesMock.data.orgs.available_organizations,
  anyTaxonomyText: 'Any Organization',
  manageTaxonomyText: 'Manage Organizations',
  anyTaxonomyURL: '/organizations/clear',
  manageTaxonomyURL: '/organizations',
  isOpen: true,
};

const fixtures = {
  rendering: { ...props },
};

describe('TaxonomyDropdown', () =>
  testComponentSnapshotsWithFixtures(TaxonomyDropdown, fixtures));
