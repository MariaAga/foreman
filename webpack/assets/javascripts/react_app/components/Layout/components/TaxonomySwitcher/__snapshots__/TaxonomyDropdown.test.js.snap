// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaxonomyDropdown rendering 1`] = `
<CustomContextSelector
  className=""
  id="organization-dropdown"
  isOpen={true}
  menuAppendTo="inline"
  onSearchButtonClick={[Function]}
  onSearchInputChange={[Function]}
  onSelect={[Function]}
  onToggle={[Function]}
  screenReaderLabel="Selected Taxonomy:"
  searchButtonAriaLabel="Search menu items"
  searchInputPlaceholder="Search"
  searchInputValue=""
  searchProps={
    Object {
      "className": "taxonomy_search",
      "id": "search_taxonomy_organization",
    }
  }
  showFilter={false}
  staticGroup={
    Object {
      "items": Array [
        Object {
          "className": "organizations_clear",
          "href": "/organizations/clear",
          "onClick": [Function],
          "title": "Any Organization",
        },
        Object {
          "className": "organization",
          "href": "/organizations",
          "title": "Manage Organizations",
        },
      ],
      "title": "Organization",
    }
  }
  toggleText="org1"
>
  <ContextSelectorItem
    className=""
    isDisabled={false}
    key="0"
    onClick={[Function]}
    sendRef={[Function]}
  >
    <a
      className="organization_menuitem"
      href="/organizations/1-org1/select"
      id="select_taxonomy_org1"
      onClick={[Function]}
    >
      org1
    </a>
  </ContextSelectorItem>
  <ContextSelectorItem
    className=""
    isDisabled={false}
    key="1"
    onClick={[Function]}
    sendRef={[Function]}
  >
    <a
      className="organization_menuitem"
      href="/organizations/2-org2/select"
      id="select_taxonomy_org2"
      onClick={[Function]}
    >
      org2
    </a>
  </ContextSelectorItem>
</CustomContextSelector>
`;
